ethereum/permissioning/src/main/java/tech/pegasys/pantheon/ethereum/permissioning/node/NodePermissioningController.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.permissioning.node;
+
+import tech.pegasys.pantheon.ethereum.permissioning.node.provider.SyncStatusNodePermissioningProvider;
+import tech.pegasys.pantheon.util.enode.EnodeURL;
+
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class NodePermissioningController {
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final Optional<SyncStatusNodePermissioningProvider> syncStatusNodePermissioningProvider;
+  private Optional<Runnable> peerDiscoveryCallback = Optional.empty();
+
+  public NodePermissioningController(
+      final SyncStatusNodePermissioningProvider syncStatusNodePermissioningProvider) {
+    this.syncStatusNodePermissioningProvider = Optional.of(syncStatusNodePermissioningProvider);
+    this.syncStatusNodePermissioningProvider.get().setHasReachedSyncCallback(this::onReachingSync);
+  }
+
+  public NodePermissioningController() {
+    this.syncStatusNodePermissioningProvider = Optional.empty();
+  }
+
+  public boolean isPermitted(final EnodeURL sourceEnode, final EnodeURL destinationEnode) {
+    LOG.trace("Checking node permission: {} -> {}", sourceEnode, destinationEnode);
+
+    return syncStatusNodePermissioningProvider
+        .map((provider) -> provider.isPermitted(sourceEnode, destinationEnode))
+        .orElse(true);
+  }
+
+  public boolean delayPeerDiscovery() {
```

Comment:
squeeze this and peerDiscoveryCallback together into 1 function that either calls the runnable immediately or delays it based on whats necessary.

Also be aware of threading issues between setting the discovery callback and onReachingSync getting called by the observable thread.

Reply:
And probably empty the callback optional before running it so you get a latch that stops anything from racing to call it twice